src/test/java/Homework21.java

```diff
@@ -0,0 +1,67 @@
+import org.openqa.selenium.By;
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import java.time.Duration;
+
+
+
+public class Homework21 extends BaseTest {
+    String playlistName = "Deleted Playlist";
+
+
+    @Test
+
+    public void renamePlaylist() throws InterruptedException {
+        String updatedPlaylistMsg = "Updated playlist \"Deleted Playlist.\"";
+
+
+        doubleClickPlaylist();
+        enterNewPlaylistName();
+
+
+        Assert.assertEquals(getRenamePlaylistSuccessMsg(), updatedPlaylistMsg);
+    }
+
+
+    private String getRenamePlaylistSuccessMsg() {
+        WebElement notification = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("div.success.show")));
+        return notification.getText();
+    }
+
+    private void enterNewPlaylistName() {
+        WebElement playlistInputField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[name='name']")));
+        playlistInputField.sendKeys(Keys.chord(Keys.CONTROL, "A"));
+        playlistInputField.sendKeys(Keys.BACK_SPACE);
+        playlistInputField.sendKeys(playlistName);
+        playlistInputField.sendKeys(Keys.ENTER);
+
+    }
+
+    private void doubleClickPlaylist() {
+        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id='playlists']/ul/li[3]")));
+        actions.doubleClick(playlistElement).perform();
+    }
+}
```

Comment:
Page Object Model (POM): recommended to associate the methods with the corresponding page class where they are relevant. In the POM, each page or section of a web application is represented by a separate class. These classes encapsulate the elements and actions specific to that page or section.

Move these methods to the corresponding class according to the POM.

Reply:
Hello, I resubmitted my hw-22, but after refactoring and doing step by step from your comments (greatly appreciate,  that it was in Russian), my code is not working and giving errors.  Can you please check and comment what I did wrong?

Reply:
Yes, sure

You got this problem, please correct it. 

![image](https://github.com/LiudmilaChernyh/A49-QA-Automation-Framework/assets/104761482/8b3f4e16-3a0e-4a76-b4f4-94994248971f)

HINT

**1)** 

```
    public void enterNewPlaylistName(String playlistName) {
        WebElement playlistInputField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[name='name']")));
        playlistInputField.sendKeys(Keys.chord(Keys.CONTROL, "A"));
        playlistInputField.sendKeys(Keys.BACK_SPACE);
        playlistInputField.sendKeys(playlistName);
        playlistInputField.sendKeys(Keys.ENTER);

    }
    public void doubleClickPlaylist() {
        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(playlist));
        actions.doubleClick(playlistElement).perform();
    }
```
![image](https://github.com/LiudmilaChernyh/A49-QA-Automation-Framework/assets/104761482/78df7e7a-fed6-497d-adb3-2543ffb70e96)

**2)**

```
    public void  login() {
        provideEmail("liudmila.chernyh@testpro.io");
        providePassword("te$t$tudent");
        clickSubmit();
    }
```

![image](https://github.com/LiudmilaChernyh/A49-QA-Automation-Framework/assets/104761482/02889d7d-f824-458d-b160-71126b9a0535)


**3)** 

```
    @BeforeMethod
    @Parameters({"BaseURL"})
    public static void launchBrowser(String BaseURL) {
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--disable-notifications","--remote-allow-origins=*", "--incognito","--start-maximized");
        options.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"});
        driver = new ChromeDriver(options);
        url=BaseURL;
        driver.get(url);

          }
```

![image](https://github.com/LiudmilaChernyh/A49-QA-Automation-Framework/assets/104761482/747e6cbd-e518-4c37-ac6d-72fe2d407791)

**4)** 

```
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import java.time.Duration;


public class BasePage {
    protected Actions actions;
    protected WebDriverWait wait;
    protected WebDriver driver;

    public BasePage(WebDriver givenDriver){
    driver =givenDriver;
    wait= new WebDriverWait(driver, Duration.ofSeconds(5));
    actions = new Actions(driver);

    }
    public  WebElement findElement (By locator){
        return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

}
```

**5)**


```
import org.testng.Assert;
import org.testng.annotations.Test;
import pages.HomePage;
import pages.LoginPage;

public class Homework21 extends BaseTest {


    @Test
    public void renamePlaylist() {
        String PlaylistName = "Deleted Playlist+";
        String updatedPlaylistMsg = "Updated playlist \"Deleted Playlist+.\"";
        LoginPage loginPage = new LoginPage(driver);
        HomePage homePage = new HomePage(driver);

        loginPage.login();
        homePage.doubleClickPlaylist();
        homePage.enterNewPlaylistName(PlaylistName);

        Assert.assertEquals(homePage.getRenamePlaylistSuccessMsg(), updatedPlaylistMsg);
    }

}
```

or download 
[A49-QA-Automation-Framework.zip](https://github.com/LiudmilaChernyh/A49-QA-Automation-Framework/files/12541983/A49-QA-Automation-Framework.zip)

Reply:
Thank you!!! It finally passed!!!